src/lib.rs

```diff
@@ -135,15 +136,13 @@ pub trait Socket {
     /// connections on it.
     ///
     /// Returns the bound-to endpoint, with the port resolved (if applicable).
-    async fn bind(&mut self, endpoint: impl TryIntoEndpoint + 'async_trait)
-        -> ZmqResult<&Endpoint>;
+    async fn bind(&mut self, endpoint: impl TryIntoEndpoint + 'async_trait) -> ZmqResult<Endpoint>;
```

Comment:
I think in an ideal world, we would never need to clone the endpoint once we convert it, so in theory the endpoint returned should be a ref, rather than owned. That being said, internally we have to clone right now in order to insert into the hashmap. But ideally this would be unnecessary.

I can take a crack at avoiding the clone, and playing around with the return value, in the same PR where I take a crack at the `binds()` return type

Reply:
IMHO this is not the most important part to work on. Basically bind/connect is called only a several times comparable to the number of call to send/recv because of this additional cloning has negligible impact on performance cause it's definitely not on a critical code path. We may come back to this some time later, however for now I mostly focus on avoid additional cloning of message contents for example. For other parts I plan to return later after actual benchmarking and finding real issues.